CONTRIBUTING.md

```diff
@@ -1,95 +1,159 @@
+A stale issue or pull request (PR) is one that has not had any activity or updates for 60 days. When an issue or PR becomes stale, it is labelled as `stale`. Normally maintainers will comment on stale issue or PR to prompt participants to take action. Then, if no additional activity occurs after 14 days, this issue or PR will be closed. If an update/comment occur on stale issues or pull requests, the stale label will be removed and the timer will restart.
 
-Some simple rules to ensure quick merge:
+## Support requests or questions
 
- - clearly point to the issue(s) you want to fix in your PR comment (e.g., `closes #12345`)
- - prefer multiple (smaller) PRs addressing individual issues over a big one trying to address multiple issues at once
- - if you need to amend your PR following comments, please squash instead of adding more commits
- - if fixing a bug or adding a feature, please add or update the relevant `CHANGELOG.md` entry with your pull request number
-   and a description of the change
+Please use slack channel for support requests or questions. Issues could be created accordingly if it is a bug or enhancement.
+
+## Contributing Workflow
+
+PR are always welcome, even if they only contain small fixes like typos or a few lines of code. If there will be a significant effort, please document it as an issue and get a discussion going before starting to work on it.
+
+Please submit a PR broken down into small changes bit by bit. A PR consisting of a lot of features and code changes may be hard to review. It is recommended to submit PRs in an incremental fashion.
+
+> **Note** If you split your pull request to small changes, please make sure any of the changes goes to main will not break anything. Otherwise, it cannot be merged until this feature complete.
+
+### Fork and clone
+
+  Fork the repository and clone the code to your local workspace.
+
+### Branch
+
+Changes should be made on your own fork in a new branch checked out from the `main` branch.
+
+### Develop, Build, and Test
+
+Write code on the new branch in your fork. Unit test cases should be added to cover the new code. For PR in [notation](https://github.com/notaryproject/notation), end-to-end (e2e) test cases should be added, see [e2e quick start](https://github.com/notaryproject/notation/blob/main/test/e2e/README.md)
+
+### Keep sync with upstream
+
+Once your branch gets out of sync with the `main` branch, please fetch and rebase from `main` branch.
+
+### Commit
+
+As Notary project repositories has integrated the [DCO (Developer Certificate of Origin)](https://github.com/apps/dco) check tool, contributors are required to sign off that they adhere to those requirements by adding a `Signed-off-by` line to the commit messages. Git has even provided a `-s` command line option to append that automatically to your commit messages, please use it when you commit your changes. For example:
+
+```shell
+git commit -s -m <commit_message>
```

Comment:
We should have both `-s` for sign-off and `-S` for signed commits. We should also link to the instructions how to set up a private key on the machine.

Reply:
+1. SSH Key is much easier to set up than GPG Key from our testing. Shall we recommend contributors use SSH key to sign commits and link to https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification#ssh-commit-signature-verification?

Reply:
Thanks @toddysm. The how-to was mentioned after the command example.

`
Notary project repositories require signed commits, please refer to [commit signatures verification](https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification) on signing commits.`

@FeynmanZhou the above guide mentioned three options. I think users can choose the best way for them.